src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +276,589 @@ create_distributed_table(PG_FUNCTION_ARGS)
+	 */
+	Oid distributionColumnType = ColumnTypeIdForRelationColumnName(relationId,
+																   distributionColumnName);
+
+	text *colocateWithTableNameText = cstring_to_text(colocateWithTableName);
+	Oid colocateWithTableId = ResolveRelationId(colocateWithTableNameText, false);
+	EnsureTableCanBeColocatedWith(relationId, replicationModel,
+								  distributionColumnType, colocateWithTableId);
+}
+
+
+/*
+ * AcquireCreateDistributedTableConcurrentlyLock does not allow concurrent create_distributed_table_concurrently
+ * operations.
+ */
+void
+AcquireCreateDistributedTableConcurrentlyLock(Oid relationId)
+{
+	LOCKTAG tag;
+	const bool sessionLock = false;
+	const bool dontWait = true;
+
+	SET_LOCKTAG_CITUS_OPERATION(tag, CITUS_CREATE_DISTRIBUTED_TABLE_CONCURRENTLY);
+
+	LockAcquireResult lockAcquired = LockAcquire(&tag, ExclusiveLock, sessionLock,
+												 dontWait);
+	if (!lockAcquired)
+	{
+		ereport(ERROR, (errmsg("another create_distributed_table_concurrently "
+							   "operation is in progress"),
+						errhint("Make sure that the concurrent operation has "
+								"finished and re-run the command")));
+	}
+}
+
+
+/*
+ * CreateDistributedTableConcurrently distributes a table by first converting
+ * it to a Citus local table and then splitting the shard of the Citus local
+ * table.
+ *
+ * If anything goes wrong during the second phase, the table is left as a
+ * Citus local table.
+ */
+static void
+CreateDistributedTableConcurrently(Oid relationId, char *distributionColumnName,
+								   char distributionMethod,
+								   char *colocateWithTableName,
+								   int shardCount,
+								   bool shardCountIsStrict)
+{
+	/*
+	 * We disallow create_distributed_table_concurrently in transaction blocks
+	 * because we cannot handle preceding writes, and we block writes at the
+	 * very end of the operation so the transaction should end immediately after.
+	 */
+	PreventInTransactionBlock(true, "create_distributed_table_concurrently");
+
+	/*
+	 * do not allow multiple create_distributed_table_concurrently in the same
+	 * transaction. We should do that check just here because concurrent local table
+	 * conversion can cause issues.
+	 */
+	ErrorIfMultipleNonblockingMoveSplitInTheSameTransaction();
+
+	/* do not allow concurrent CreateDistributedTableConcurrently operations */
+	AcquireCreateDistributedTableConcurrentlyLock(relationId);
+
+	if (distributionMethod != DISTRIBUTE_BY_HASH)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("only hash-distributed tables can be distributed "
+							   "without blocking writes")));
+	}
+
+	if (ShardReplicationFactor > 1)
+	{
+		ereport(ERROR, (errmsg("cannot distribute a table concurrently when "
+							   "citus.shard_replication_factor > 1")));
+	}
+
+	EnsureCoordinatorIsInMetadata();
+	EnsureCitusTableCanBeCreated(relationId);
+
+	EnsureValidDistributionColumnNoLock(relationId, distributionColumnName);
```

Comment:
minor: the "NoLock" in the name feels a bit pedantic in this case. Does it mean other Ensure functions do aquire a lock?